Fixed routing issue caused by JS state management

tehuanmelo opened the pull request.

tehuanmelo committed:
fix(browser): resolve issue with state history not saving 404 pages

- Fixed a bug where the browser was not saving the state of 404 pages in the history.

tehuanmelo committed:
fix(popstate): ensure state and URL update correctly with page content

- Fixed an issue where the URL and state were not being updated correctly on popstate events.
- Now, the state and URL reflect the current content of the page as intended.

Taanviir reviewed (changes requested):
I have looked at the code and tested the changes however there was an issue where any link other than `localhost` threw a 404 error if entered in the URL search bar.

Taanviir commented on backend/static/scripts/main.js:
```diff
@@ -50,36 +51,39 @@ function handleSpaLinkEvent(target) {
+    }
+  
     const newTitle = page.querySelector('title').innerHTML;
+    document.title = '';
```
also why is this being done

tehuanmelo replied:
it must be there, to prevent caching

tehuanmelo committed:
fix(router): resolve issue preventing users from manually inputting URLs

- Fixed a bug where users were unable to manually input URLs and navigate to the correct page.

tehuanmelo commented:
> I have looked at the code and tested the changes however there was an issue where any link other than `localhost` threw a 404 error if entered in the URL search bar.

Issue fixed.
Please review it and let me know if the error persists.

Taanviir reviewed (approved):
Everything looks good, you can go ahead and merge
